Emby.Server.Implementations/Data/BaseSqliteRepository.cs

```diff
@@ -27,123 +27,107 @@ protected BaseSqliteRepository(ILogger logger)
+                                                DefaultConnectionFlags | ConnectionFlags.Create | ConnectionFlags.ReadWrite,
+                                                null);
                 }
 
-                var queries = new List<string>
+                if (string.IsNullOrWhiteSpace(_defaultWal))
                 {
-                    //"PRAGMA cache size=-10000"
-                    //"PRAGMA read_uncommitted = true",
-                    //"PRAGMA synchronous=Normal"
-                };
+                    _defaultWal = WriteConnection.Query("PRAGMA journal_mode").SelectScalarString().First();
 
-                if (CacheSize.HasValue)
-                {
-                    queries.Add("PRAGMA cache_size=" + CacheSize.Value.ToString(CultureInfo.InvariantCulture));
+                    Logger.LogInformation("Default journal_mode for {0} is {1}", DbFilePath, _defaultWal);
                 }
 
                 if (EnableTempStoreMemory)
                 {
-                    queries.Add("PRAGMA temp_store = memory");
+                    WriteConnection.Execute("PRAGMA temp_store = memory");
                 }
                 else
                 {
-                    queries.Add("PRAGMA temp_store = file");
-                }
-
-                foreach (var query in queries)
-                {
-                    db.Execute(query);
+                    WriteConnection.Execute("PRAGMA temp_store = file");
                 }
             }
             catch
             {
-                using (db)
-                {
-
-                }
 
                 throw;
             }
+            finally
+            {
+                WriteLock.Release();
+            }
 
-            return db;
+            // Add one reading connection for each thread
+            int threads = System.Environment.ProcessorCount;
+            for (int i = 0; i <= threads; i++)
+            {
+                ReadConnectionPool.Add(SQLite3.Open(DbFilePath, DefaultConnectionFlags | ConnectionFlags.ReadOnly, null));
+            }
         }
 
-        public IStatement PrepareStatement(SQLiteDatabaseConnection connection, string sql)
+        protected ManagedConnection GetConnection(bool isReadOnly = false)
         {
-            return connection.PrepareStatement(sql);
-        }
+            if (isReadOnly)
+            {
+                return new ManagedConnection(ReadConnectionPool.Take(), ReadConnectionPool);
+            }
+            else
+            {
+                if (WriteConnection == null)
+                {
+                    throw new InvalidOperationException("Can't access the write connection at this time.");
+                }
 
-        public IStatement PrepareStatementSafe(SQLiteDatabaseConnection connection, string sql)
-        {
-            return connection.PrepareStatement(sql);
+                WriteLock.Wait();
+                return new ManagedConnection(WriteConnection, WriteLock);
```

Comment:
why cannot `ManagedConnection` take this lock? it's obviously going to be releasing it, so why it cannot acquire it, too?

Reply:
IMO that's not something you should do in a constructor.